File: Pandacap.LowLevel/Data/TheirContent/ActivityPub/InboxActivityStreamsPost.fs
namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations
open Pandacap.Html
open Pandacap.PlatformBadges

type InboxActivityStreamsUser() =
    member val Id = "" with get, set
    member val Username = nullString with get, set
    member val Usericon = nullString with get, set

type InboxActivityStreamsImage() =
    member val Url = "" with get, set
    member val Name = nullString with get, set

type InboxActivityStreamsPost() =
    [<Key>]
    member val Id = Guid.Empty with get, set

    member val AnnounceId = nullString with get, set
    member val ObjectId = "" with get, set
    member val Author = new InboxActivityStreamsUser() with get, set
    member val PostedBy = new InboxActivityStreamsUser() with get, set
    member val PostedAt = DateTimeOffset.MinValue with get, set
    member val Summary = nullString with get, set
    member val Sensitive = false with get, set
    member val Name = nullString with get, set
    member val Content = nullString with get, set
    member val Attachments = new ResizeArray<InboxActivityStreamsImage>() with get, set
    member val DismissedAt = nullDateTimeOffset with get, set

    member this.TextContent = TextConverter.FromHtml this.Content

    interface IInboxPost with
        member this.DismissedAt
            with get () = this.DismissedAt
             and set value = this.DismissedAt <- value
        member _.IsPodcast = false
        member this.IsShare = this.PostedBy.Id <> this.Author.Id

    interface IPost with
        member _.Platform = ActivityPub
        member this.Url = this.ObjectId
        member this.DisplayTitle = ExcerptGenerator.FromFirst 60 (seq {
            this.Name
            this.Summary
            TextConverter.FromHtml this.Content
        })
        member this.Id = $"{this.Id}"
        member this.LinkUrl = $"/RemotePosts?id={Uri.EscapeDataString(this.ObjectId)}"
        member this.PostedAt = this.PostedAt
        member this.ProfileUrl = this.PostedBy.Id
        member this.Thumbnails = [
            if not this.Sensitive then
                for a in this.Attachments do {
                    new IPostThumbnail with
                        member _.AltText = a.Name
                        member _.Url = a.Url
                }
        ]
        member this.Username = this.PostedBy.Username
        member this.Usericon = this.PostedBy.Usericon

File: Pandacap.Database/InboxActivityStreamsPost.cs
using Pandacap.Text;
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;
using System.ComponentModel.DataAnnotations.Schema;

namespace Pandacap.Database
{
    public class InboxActivityStreamsPost : IInboxPost
    {
        public Guid Id { get; set; }

        public string? AnnounceId { get; set; }

        public string ObjectId { get; set; } = "";

        public class User
        {
            public string Id { get; set; } = "";

            public string? Username { get; set; }

            public string? Usericon { get; set; }
        }

        public User Author { get; set; } = new();

        public User PostedBy { get; set; } = new();

        public DateTimeOffset PostedAt { get; set; }

        public string? Summary { get; set; }

        public bool Sensitive { get; set; }

        public string? Name { get; set; }

        public string? Content { get; set; }

        public class Image : IPostThumbnail
        {
            public string Url { get; set; } = "";

            public string? Name { get; set; }

            string IPostThumbnail.AltText => Name ?? "";
        }

        public List<Image> Attachments { get; set; } = [];

        public DateTimeOffset? DismissedAt { get; set; }

        [NotMapped]
        public string TextContent => TextConverter.FromHtml(Content);

        bool IInboxPost.IsPodcast => false;

        bool IInboxPost.IsShare => PostedBy.Id != Author.Id;

        Badge IPost.Badge => Badges.ActivityPub.WithHostFromUriString(ObjectId);

        string IPost.DisplayTitle => Summary ?? Name ?? ExcerptGenerator.FromText(60, TextContent);

        string IPost.Id => $"{Id}";

        string IPost.InternalUrl => $"/RemotePosts?id={Uri.EscapeDataString(ObjectId)}";

        string IPost.ExternalUrl => ObjectId;

        string? IPost.ProfileUrl => PostedBy.Id;

        IEnumerable<IPostThumbnail> IPost.Thumbnails => Attachments;

        string? IPost.Username => PostedBy.Username;

        string? IPost.Usericon => PostedBy.Usericon;
    }
}
